### PR TITLE
Support "tianma-499" panel in motorola-harpia

### PR DESCRIPTION
**Pull request by Sorixelle**
Display tested and verified functioning correctly on real hardware.

Few things to possibly look at in future:
- "tianma-499" probably isn't the correct name for the panel - need to figure out what it actually is (somehow)
- Display regulators are currently implemented as regulator-fixed in the device tree. Should look into somehow making the tps65132 driver work with this regulator, or write a whole new nt50358 driver based on tps65132.

### CODE REVIEW

File: arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts [hunk 1]
```diff
@@ -121,6 +137,76 @@
 		pinctrl-0 = <&usb_id_default>;
 		pinctrl-1 = <&usb_id_sleep>;
 	};
+
+	/*
+	 * harpia uses a NT50358 regulator to drive the display panel. This
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts]: Minor nitpick: Try to wrap lines in a sensible way. If only one word fits, prefer adding the line break earlier (and start the new sentence on the next line).

File: drivers/gpu/drm/panel/Kconfig [hunk 2]
```diff
@@ -357,6 +357,11 @@ config DRM_PANEL_SONY_ACX565AKM
 	  Say Y here if you want to enable support for the Sony ACX565AKM
 	  800x600 3.5" panel (found on the Nokia N900).
 
+config DRM_PANEL_TIANMA_499
+	tristate "Tianma 499 panel"
+	depends on GPIOLIB && OF && SPI
```
Comment (Minecrell) [inline drivers/gpu/drm/panel/Kconfig]: You don't depend on SPI, but you probably want to select `DRM_MIPI_DSI`

File: arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts [hunk 3]
```diff
@@ -121,6 +137,76 @@
...
+	 * kernel.
+	 * (https://github.com/MotorolaMobilityLLC/kernel-msm/blob/nougat-7.1.1-release-harpia/arch/arm/boot/dts/qcom/msm8916-harpia.dtsi#L202)
+	 * TODO: figure out what needs to be done to get a proper driver working
+	 * for this regulator. For now, we can just use fixed-regulator, as the
+	 * regulator reads voltages on startup from an onboard EEPROM, and we
+	 * don't need to adjust the regulator's values.
+	 */
+	nt50358_pos: nt50358-pos {
+		compatible = "regulator-fixed";
+		regulator-name = "nt50358_pos";
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts]: Is there a reason why you prefer `_` over `-` here? I guess I don't mind but I would have used

```suggestion
		regulator-name = "nt50358-pos";
```

File: arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts [hunk 4]
```diff
@@ -121,6 +137,76 @@
...
+	pinctrl-1 = <&pmx_mdss_sleep>;
+
+	panel@0 {
+		compatible = "tianma,499";
+		reg = <0>;
+
+		vsp-supply = <&nt50358_pos>;
+		vsn-supply = <&nt50358_neg>;
+		reset-gpios = <&msmgpio 25 GPIO_ACTIVE_HIGH>;
+		backlight-gpios = <&msmgpio 37 GPIO_ACTIVE_HIGH>;
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts]: Can you add pinctrl for this GPIO? probably just bias-disable, drive-strength 2 as usual

File: drivers/gpu/drm/panel/panel-tianma-499.c [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+	struct mipi_dsi_device *dsi = bl_get_data(bl);
+	struct tianma_499 *ctx = mipi_dsi_get_drvdata(dsi);
+	u16 brightness = bl->props.brightness;
+
+	if (bl->props.power != FB_BLANK_UNBLANK ||
+	    bl->props.fb_blank != FB_BLANK_UNBLANK ||
+	    bl->props.state & (BL_CORE_SUSPENDED | BL_CORE_FBBLANK))
+		brightness = 0;
+
+	dev_err(&dsi->dev, "Updating GPIO: %d\n", !!brightness);
```
Comment (Minecrell) [inline drivers/gpu/drm/panel/panel-tianma-499.c]: remove

File: arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts (hunk 3 above)
Comment (Sorixelle) [reply arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts]: No particular reason, I'll get that changed.

File: arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts [hunk 6]
```diff
@@ -264,6 +353,61 @@
...
+			function = "gpio";
+			pins = "gpio25";
+		};
+		pinconf {
+			drive-strength = <2>;
+			bias-pull-down;
+		};
+	};
+
+	panel_backlight: panel_backlight {
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-harpia.dts]: Let's try to order pinctrl alphabetically

File: drivers/gpu/drm/panel/panel-tianma-499.c [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+	int ret;
+
+	ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	ctx->supplies[0].supply = "vsp";
+	ctx->supplies[1].supply = "vsn";
+	ret = devm_regulator_bulk_get(dev, ARRAY_SIZE(ctx->supplies),
+				     ctx->supplies);
```
Comment (Minecrell) [inline drivers/gpu/drm/panel/panel-tianma-499.c]: somehow a space got lost here for proper alignment

```
CHECK: Alignment should match open parenthesis
#350: FILE: drivers/gpu/drm/panel/panel-tianma-499.c:303:
+	ret = devm_regulator_bulk_get(dev, ARRAY_SIZE(ctx->supplies),
+				     ctx->supplies);
```

```suggestion
				      ctx->supplies);
```

**Pull request merged**